### ISSUE TITLE
TF_GraphImportGraphDef breaks down while importing graphs with control flow ops

### ISSUE THREAD

**Issue by juraj-bicikl** [tensorflow/tensorflow#5406]
I've encountered issues when trying to load graphs from python which contain
dropout or batch norm. In the following example:
```python
    x = tf.placeholder(tf.float64, [None, 3])

    w = tf.Variable(tf.random_normal([3, 3], dtype=tf.float64))
    h = tf.matmul(x, w)

    is_training = tf.Variable(True, trainable=False)
    h_dropout = tf.contrib.layers.dropout(h, keep_prob=0.5, is_training=is_training)
```
TF_GraphImportGraphDef breaks down and I get the following message:

*SIGSEGV (0xb) at pc=0x000000013539dcac, pid=23288, tid=0x0000000000006a1b
C  [libtensorflow.dylib+0x14cac]  tensorflow::shape_inference::InferenceContext::Rank(tensorflow::shape_inference::ShapeHandle) const+0x1c*

The same happens if I replace dropout operation with batch norm:
```python
tf.contrib.layers.batch_norm(h, center=True, scale=True, is_training=is_training)
```

It seems to me this problem is related to the presence of control flow ops.

**Comment by aselle** [comment]
It is very difficult to help you when you do not provide basic information about your environment. What OS? What Version of TensorFlow? Also, you should create a simple test example that can be reproduced. Thanks!

**Comment by juraj-bicikl** [comment]
I am using Mac OS X 10.11.6, and Python 3.5.1 with tensorflow 0.11.0rc0 version.
I compiled the following C++ code on top of 0.11.0rc0 version (commit 2dea76c).

Python code where I create a graph and serialize it as GraphDef
```python
import tensorflow as tf
import numpy as np
import sys

graph = tf.get_default_graph()

x = tf.placeholder(tf.float64, [None, 3])
w = tf.Variable(tf.random_normal([3, 3], dtype=tf.float64))
h = tf.matmul(x, w)

is_training = tf.Variable(True, trainable=False)
h_dropout = tf.contrib.layers.dropout(h, keep_prob=0.5, is_training=is_training)

sys.stdout.buffer.write(graph.as_graph_def().SerializeToString())
```

C++ code where I import the graph:
```C++
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <fstream>
#include <string>
#include <iostream>
#include "tensorflow/c/c_api.h"
#include "tensorflow/core/framework/graph.pb.h"
#include "tensorflow/core/framework/graph_def_util.h"

int main(int argc, char* argv[]) {

    // loading graph proto
    tensorflow::GraphDef graph_def;
    std::ifstream f_in(argv[1]);
    bool ctrl = graph_def.ParseFromIstream(&f_in);
    f_in.close();
    if (ctrl) std::cout << "deserialization successful\n\n";

    std::cout << "Graph contents:\n";
    std::cout << tensorflow::SummarizeGraphDef(graph_def) << "\n";

    // loading graph_def as TF_Graph
    TF_ImportGraphDefOptions* opts = TF_NewImportGraphDefOptions();
    TF_Status* status = TF_NewStatus();
    TF_Graph* tf_graph = TF_NewGraph();
    std::string gd_as_string = std::string();
    graph_def.SerializeToString(&gd_as_string);
    TF_Buffer* gd_as_buffer = TF_NewBufferFromString(gd_as_string.c_str(), gd_as_string.length());
    TF_GraphImportGraphDef(tf_graph, gd_as_buffer, opts, status);

    if (TF_GetCode(status) == 0)
      std::cout << "graph_def succesfully loaded into tf_graph\n";
    else
      std::cout << TF_Message(status) << '\n';

    TF_DeleteImportGraphDefOptions(opts);
    TF_DeleteStatus(status);
    TF_DeleteGraph(tf_graph);

    return 0;
}
```
When I run the C++ program and try to load GraphDef generated by the python code (which 
includes dropout), I get:

*Segmentation fault: 11*

If I do not include the dropout layer, everything works fine.

**Comment by asimshankar** [comment]
The segfault was fixed in https://github.com/tensorflow/tensorflow/commit/7ea2f7d2689d9686c93650bf5bcc4c8ba459377d
which unfortunately isn't part of the 0.11 release, but will be part of 0.12.

I'm going to go ahead and close this for now since the problem mentioned in the issue is fixed at head. @juraj-bicikl - hopefully this is okay and you don't _require_ this in 0.11.

( @cwhipkey  FYI)

**Comment by juraj-bicikl** [comment]
Thanks a lot!
